### PR TITLE
Added functionality to update Users' thumbnail photos

### PR DESCRIPTION
**Pull request by calvli**
Added functionality to update Users' thumbnail photos
Updated gitignore to ignore config file
Updated README

### CODE REVIEW

**Comment by calvli** [comment]
@AnaghaKothurwar is added to the review. #Closed

File: WebAppGraphAPI/Web.config [hunk 1]
```diff
@@ -9,11 +9,15 @@
...
+      
+    <!-- Modify these values-->
     <add key="ida:ClientId" value="c7106273-7c25-48e8-954e-cf207b5eb3c9" />
     <add key="ida:AppKey" value="wDckc8ebbawALFLeRvvxdG7zGf6N36T/JPm02oOkZsI=" />
-    <add key="ida:AADInstance" value="https://login.windows.net/{0}" />
     <add key="ida:Tenant" value="graphDir1.onMicrosoft.com" />
     <add key="ida:TenantId" value="4fd2b2f2-ea27-4fe5-a8f3-7b1a7c975f34" />
+    <!-- Modify these values-->
+
+    <add key="ida:AADInstance" value="https://login.windows.net/{0}" />
```
Comment (dkershaw10) [inline WebAppGraphAPI/Web.config]: We should be using https://login.microsoftonline.com.  login.windows.net is old, and while it works, it introduces an extra hop.

File: WebAppGraphAPI/Web.config [hunk 2]
```diff
@@ -9,11 +9,15 @@
...
     <add key="ida:PostLogoutRedirectUri" value="http://localhost:44322/" />
     <add key="ida:GraphApiVersion" value="2013-11-08" />
```
Comment (dkershaw10) [inline WebAppGraphAPI/Web.config]: This should be the latest version - i.e. "1.6"

File: WebAppGraphAPI/Web.config [hunk 3]
```diff
@@ -9,11 +9,15 @@
     <add key="webpages:Enabled" value="false" />
     <add key="ClientValidationEnabled" value="true" />
     <add key="UnobtrusiveJavaScriptEnabled" value="true" />
+      
+    <!-- Modify these values-->
     <add key="ida:ClientId" value="c7106273-7c25-48e8-954e-cf207b5eb3c9" />
     <add key="ida:AppKey" value="wDckc8ebbawALFLeRvvxdG7zGf6N36T/JPm02oOkZsI=" />
```
Comment (dkershaw10) [inline WebAppGraphAPI/Web.config]: Does the app work with the default values?

File: README.md [hunk 4]
```diff
@@ -35,57 +35,47 @@ The sample app is preconfigured to read data from a Demonstration company (Graph
...
-2. In `web.config` add this line in the `<system.web>` section: `<sessionState timeout="525600" />`.  This increases the ASP.Net session state timeout to it's maximum value so that access tokens and refresh tokens cache in session state aren't cleared after the default timeout of 20 minutes.
-3. Build and run your application - you will need to authenticate with valid user credentials for your company when you run the application.
-
-Note: there is an issues with uploading jpeg file images for user's thumbnail photos when using this app
-from Chrome or FireFox (Internet Explorer works) - we will debug this in more detail and update when this issue
-is addressed.
+2. Find your tenantID. Your tenantId can be discovered by opening the following metadata.xml document: https://login.windows.net/GraphDir1.onmicrosoft.com/FederationMetadata/2007-06/FederationMetadata.xml - replace "graphDir1.onMicrosoft.com", with your tenant's domain value (any domain that is owned by the tenant will work). The tenantId is a guid, that is part of the sts URL, returned in the first xml node's sts url ("EntityDescriptor"): e.g. "https://sts.windows.net/".
```
Comment (dkershaw10) [inline README.md]: Change login.windows.net -> login.microsoftonline.com

File: README.md [hunk 5]
```diff
@@ -35,57 +35,47 @@ The sample app is preconfigured to read data from a Demonstration company (Graph
 
 NOTE: If you already have a user account in your Azure Active Directory tenant, you can skip to the next step.  This sample will not work with a Microsoft account, so if you signed in to the Azure portal with a Microsoft account and have never created a user account in your directory before, you need to do that now.  If you create an account and want to use it to sign-in to the Azure portal, don't forget to add the user account as a co-administrator of your Azure subscription.
```
Comment (dkershaw10) [inline README.md]: If you are getting the user to use the new Azure Portal (which is a great idea) then your last line is incorrect.  AFAIK an Azure subscription (or co-admin of the subscription) is not required to use the new Azure Portal.  You just need to be a member of a tenant (even is you sign in with an MSA - as long as that MSA is a member of one or more tenants).

File: README.md [hunk 6]
```diff
@@ -35,57 +35,47 @@ The sample app is preconfigured to read data from a Demonstration company (Graph
...
-
-NOTE: the permission "Access your organization's directory" allows the application to access your organization's directory on behalf of the signed-in user - this is a delegation permission and must be consented by the Administrator for web apps (such as this demo app).
-The permission "Enable sign-on and read users' profiles" allows users to sign in to the application with their organizational accounts and lets the application read the profiles of signed-in users, such as their email address and contact information - this is a delegation permission, and can be consented to by the user.
-The other permissions, "Read Directory data" and "Read and write Directory data", are Delegation and Application Permissions, which only the Administrator can grant consent to.
+1. Sign in to the [Azure portal](https://portal.azure.com).
+2. On the top right, click on your account and under the **Directory** list, choose an Active Directory tenant where you have admin permissions.
+3. Click on **More Services** in the left hand nav, and choose **Azure Active Directory**.
```
Comment (dkershaw10) [inline README.md]: Simpler to type "App registrations" in the search filter and then click "App registrations"

File: README.md [hunk 7]
```diff
@@ -35,57 +35,47 @@ The sample app is preconfigured to read data from a Demonstration company (Graph
...
+7. Find the Application ID value and copy it to the clipboard.
+8. In the Reply URL, add the reply URL address used to return the authorization code returned during Authorization code flow, eg https://localhost:44322/". **NOTE:** If you see TLS error messages, try changing your reply URL to use http instead of http*s*.
+9. From the Settings menu, choose **Keys** and add a key - select a key duration of either 1 year or 2 years. When you save this page, the key value will be displayed, copy and save the value in a safe location - you will need this key later to configure the project in Visual Studio - this key value will not be displayed again, nor retrievable by any other means, so please record it as soon as it is visible from the Azure Portal.
+10. Configure Permissions for your application - in the Settings menu, choose the 'Required permissions' section, click on **Add**, then **Select an API**, and select 'Microsoft Graph' (this is the Graph API). Then, click on  **Select Permissions** and select 'Access the directory as the signed-in user' and 'Sign in and read user profile'.
```
Comment (dkershaw10) [inline README.md]: Ah no.  Microsoft Graph is the **new** graph that we will want people using.  However, this is AAD Graph and the client library is for AAD Graph.  You need to have them select "Windows Azure Active Directory"

File: README.md [hunk 8]
```diff
@@ -35,57 +35,47 @@ The sample app is preconfigured to read data from a Demonstration company (Graph
...
-from Chrome or FireFox (Internet Explorer works) - we will debug this in more detail and update when this issue
-is addressed.
+2. Find your tenantID. Your tenantId can be discovered by opening the following metadata.xml document: https://login.windows.net/GraphDir1.onmicrosoft.com/FederationMetadata/2007-06/FederationMetadata.xml - replace "graphDir1.onMicrosoft.com", with your tenant's domain value (any domain that is owned by the tenant will work). The tenantId is a guid, that is part of the sts URL, returned in the first xml node's sts url ("EntityDescriptor"): e.g. "https://sts.windows.net/".
+3. In `web.config` add this line in the `<system.web>` section: `<sessionState timeout="525600" />`.  This increases the ASP.Net session state timeout to it's maximum value so that access tokens and refresh tokens cache in session state aren't cleared after the default timeout of 20 minutes.
+4. Build and run your application - you will need to authenticate with valid user credentials for your company when you run the application. **Note:** you might have to manually create a packages folder for NuGet.
```
Comment (dkershaw10) [inline README.md]: This final note seems a little out of place???

File: WebAppGraphAPI/Controllers/UsersController.cs [hunk 9]
```diff
@@ -254,17 +254,24 @@ public class UsersController : Controller
         ///     Processes editing of an existing <see cref="User" />.
         /// </summary>
         /// <param name="user"><see cref="User" /> to be edited.</param>
+        /// <param name="values">new values user, taken from form</param>
```
Comment (alawon) [inline WebAppGraphAPI/Controllers/UsersController.cs]: Can we reword this `FormCollection values` parameter description? I might be missing some context but it's not immediately clear to me, particularly the "`new values user`" part.

File: WebAppGraphAPI/Controllers/UsersController.cs [hunk 10]
```diff
@@ -254,17 +254,24 @@ public class UsersController : Controller
...
             {
                 ActiveDirectoryClient client = AuthenticationHelper.GetActiveDirectoryClient();
-                IUser toUpdate = await client.Users.GetByObjectId(user.ObjectId).ExecuteAsync();
+                string userId = user.ObjectId ?? RouteData.Values["id"].ToString();
+                IUser toUpdate = await client.Users.GetByObjectId(userId).ExecuteAsync();
                 Helper.CopyUpdatedValues(toUpdate, user, values);
                 await toUpdate.UpdateAsync();
+                if (photoFile.ContentLength > 0)
+                {
+                    await toUpdate.ThumbnailPhoto.UploadAsync(photoFile.InputStream, "application/image");
```
Comment (alawon) [inline WebAppGraphAPI/Controllers/UsersController.cs]: What's the contract on `UploadAsync` (as well as the other async calls above)? Is there validation on whether the calls are successful?

File: WebAppGraphAPI/Web.config [hunk 11]
```diff
@@ -9,11 +9,15 @@
...
-    <add key="ida:AADInstance" value="https://login.windows.net/{0}" />
     <add key="ida:Tenant" value="graphDir1.onMicrosoft.com" />
     <add key="ida:TenantId" value="4fd2b2f2-ea27-4fe5-a8f3-7b1a7c975f34" />
+    <!-- Modify these values-->
```
Comment (AnaghaKothurwar) [inline WebAppGraphAPI/Web.config]: >Modify these values [](start = 9, length = 19)

I believe this the end comment tag for "modify these values" - which is not clear. Instead rearrange the values to have all constants at the top and then modifiable values at the end.

File: WebAppGraphAPI/Controllers/UsersController.cs [hunk 12]
```diff
@@ -254,17 +254,24 @@ public class UsersController : Controller
         ///     Processes editing of an existing <see cref="User" />.
         /// </summary>
         /// <param name="user"><see cref="User" /> to be edited.</param>
+        /// <param name="values">new values user, taken from form</param>
+        /// <param name="photoFile">thumbnail photo file</param>
         /// <returns>A view with list of all <see cref="User" /> objects.</returns>
         [HttpPost]
         public async Task<ActionResult> Edit(
-            User user, FormCollection values)
+            User user, FormCollection values, HttpPostedFileBase photoFile)
```
Comment (AnaghaKothurwar) [inline WebAppGraphAPI/Controllers/UsersController.cs]: >photoFile [](start = 65, length = 9)

Can photoFile be null? should there be null check before using it?

File: WebAppGraphAPI/Controllers/UsersController.cs [hunk 13]
```diff
@@ -254,17 +254,24 @@ public class UsersController : Controller
...
         {
             try
             {
                 ActiveDirectoryClient client = AuthenticationHelper.GetActiveDirectoryClient();
-                IUser toUpdate = await client.Users.GetByObjectId(user.ObjectId).ExecuteAsync();
+                string userId = user.ObjectId ?? RouteData.Values["id"].ToString();
```
Comment (AnaghaKothurwar) [inline WebAppGraphAPI/Controllers/UsersController.cs]: >RouteData.Values["id"] [](start = 49, length = 22)

Can we always use RouteData.Values["id"] instead? what are the cases when user.ObjectId will be null/non nulll?

File: WebAppGraphAPI/Web.config (hunk 1 above)
Comment (calvli) [inline WebAppGraphAPI/Web.config]: changed

---
In reply to: [80820901](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80820901) [](ancestors = 80820901)

File: WebAppGraphAPI/Web.config (hunk 2 above)
Comment (calvli) [inline WebAppGraphAPI/Web.config]: done

---
In reply to: [80820945](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80820945) [](ancestors = 80820945)

File: WebAppGraphAPI/Web.config (hunk 3 above)
Comment (calvli) [inline WebAppGraphAPI/Web.config]: Just tried it, I get **System.IdentityModel.Tokens.SecurityTokenSignatureKeyNotFoundException: IDX10500: Signature validation failed. Unable to resolve SecurityKeyIdentifier:** when using the demo account.

---
In reply to: [80820995](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80820995) [](ancestors = 80820995)

File: README.md (hunk 4 above)
Comment (calvli) [inline README.md]: done

---
In reply to: [80821186](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80821186) [](ancestors = 80821186)

File: README.md (hunk 5 above)
Comment (calvli) [inline README.md]: This line was there before, so I'm not sure what the original purpose was. Sounds like we can remove it.

---
In reply to: [80821664](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80821664) [](ancestors = 80821664)

File: README.md (hunk 6 above)
Comment (calvli) [inline README.md]: updated

---
In reply to: [80821967](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80821967) [](ancestors = 80821967)

File: README.md (hunk 8 above)
Comment (calvli) [inline README.md]: I had to do that to get the aoo to build

---
In reply to: [80822624](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80822624) [](ancestors = 80822624)

File: WebAppGraphAPI/Controllers/UsersController.cs (hunk 9 above)
Comment (calvli) [inline WebAppGraphAPI/Controllers/UsersController.cs]: Reworded it. It's just the contents in the fields of the "Edit" form

---
In reply to: [80965030](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80965030) [](ancestors = 80965030)

File: WebAppGraphAPI/Controllers/UsersController.cs (hunk 12 above)
Comment (calvli) [inline WebAppGraphAPI/Controllers/UsersController.cs]: Yeah, adding a null check

---
In reply to: [80968739](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80968739) [](ancestors = 80968739)

File: WebAppGraphAPI/Controllers/UsersController.cs (hunk 10 above)
Comment (calvli) [inline WebAppGraphAPI/Controllers/UsersController.cs]: It throws exceptions if not successful, which we catch and display on the page.

---
In reply to: [80967154](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80967154) [](ancestors = 80967154)

File: WebAppGraphAPI/Web.config (hunk 11 above)
Comment (calvli) [inline WebAppGraphAPI/Web.config]: Ok, moved the section to the end of appSettings.

---
In reply to: [80967811](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80967811) [](ancestors = 80967811)

File: WebAppGraphAPI/Controllers/UsersController.cs (hunk 13 above)
Comment (calvli) [inline WebAppGraphAPI/Controllers/UsersController.cs]: Never managed to figure out why user.ObjectId was null. Removing it.

---
In reply to: [80968869](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-web/pull/39#discussion_r80968869) [](ancestors = 80968869)

**Pull request merged**